pkg/archive/archive.go

```diff
@@ -225,6 +225,61 @@ func CompressStream(dest io.Writer, compression Compression) (io.WriteCloser, er
 	}
 }
 
+// ReplaceFileTarWrapper converts inputTarStream to a new tar stream
+// while replacing a single file called header.Name with new contents.
+// If the file with header.Name does not exist it is added to the tar stream.
+// TODO: make this into a generic tar conversion function with walkFn argument
+func ReplaceFileTarWrapper(inputTarStream io.ReadCloser, header tar.Header, file []byte) io.ReadCloser {
+	pipeReader, pipeWriter := io.Pipe()
+
+	go func() {
+		tarReader := tar.NewReader(inputTarStream)
+		tarWriter := tar.NewWriter(pipeWriter)
+
+		defer inputTarStream.Close()
+
+		for {
+			hdr, err := tarReader.Next()
+			if err == io.EOF {
+				header.Size = int64(len(file))
+
+				if err := tarWriter.WriteHeader(&header); err != nil {
```

Comment:
Instead of always writing at end, write when `err == io.EOF` or as soon as `header.Name <= hdr.Name`